### ISSUE TITLE
CoreML model error: The size of the output layer '740' in the neural network does not match the number of classes in the classifier

### ISSUE THREAD

**Issue by Kaisershmarren** [ultralytics/yolov5#4338]
I've exported the yolov5s.pt model to CoreML creating yolov5s.mlmodel.
I've followed the instructions reported in #251 and #3238.
The command I used is the following:

python3 export.py --weights yolov5s.pt --img 640 --batch 1 --train

The conversion has been ok (or at least it's seemed) and created the model (with the right class label).
![image](https://user-images.githubusercontent.com/3902189/128609746-68993c2e-45a3-40bf-aac1-62bf2492adfe.png)

 I've included it in the Xcode project provided by Apple for sampling object detection
https://developer.apple.com/documentation/vision/recognizing_objects_in_live_capture
![image](https://user-images.githubusercontent.com/3902189/128609787-9162b6a7-c7d3-4daa-b07a-e64947481c57.png)

But when the project runs on the iPhone, it throws the following error and nothing is recognised on the iPhone screen.
NSLocalizedDescription=The size of the output layer '740' in the neural network does not match the number of classes in the classifier.
![image](https://user-images.githubusercontent.com/3902189/128610083-51d68134-f86e-40f7-8851-d71dd1abd470.png)

I've also done a check using the YOLOv3 MLModel provided by Apple in its site but it works perfectly:

![image](https://user-images.githubusercontent.com/3902189/128609708-cd572c0e-bd1b-44a4-85fc-3b17cfffe4b3.png)

My environment is:

MacOS Big Sur 11.5.1
Python 3.9
torch 1.9.0 CPU
onnx 1.10.0
scikit-learn version 0.24.2

**Comment by Kaisershmarren** [comment]
Unfortunately no... But I've just realised I didn't write all the steps I've done.
I was trying to solve the problem of absence of classes labels  in the iOS coreml model, and I've modified the export.py. 
I managed to have the classes labels but it came out the error on "740"...

**Comment by nathan-gage** [comment]
I can't even get the class labels working--how did you do that?

**Comment by MaxNazarov93** [comment]
@ngagesmu you should create ClassifierConfig with your own class labels and pass it into convert function
```
from coremltools.converters import ClassifierConfig

classifier_config = ClassifierConfig(class_labels=['class1', 'class2', 'class3', 'class4'])
ct_model = ct.convert(ts, inputs=[ct.ImageType('image', shape=im.shape, scale=1 / 255, bias=[0, 0, 0])], classifier_config=classifier_config)
```

**Comment by zjsjayce** [comment]
> @ngagesmu you should create ClassifierConfig with your own class labels and pass it into convert function
> 
> ```
> from coremltools.converters import ClassifierConfig
> 
> classifier_config = ClassifierConfig(class_labels=['class1', 'class2', 'class3', 'class4'])
> ct_model = ct.convert(ts, inputs=[ct.ImageType('image', shape=im.shape, scale=1 / 255, bias=[0, 0, 0])], classifier_config=classifier_config)
> ```
@MaxNazarov93  I added the labels as your suggestion, but it also doesn't work.
"NSLocalizedDescription = "The size of the output layer 'var_1162' in the neural network does not match the number of classes in the classifier.";
}"

**Comment by zjsjayce** [comment]
> 我已将 yolov5s.pt 模型导出到 CoreML 创建 yolov5s.mlmodel。 我已按照#251和#3238中报告的说明进行操作。 我使用的命令如下：
> 
> python3 export.py --weights yolov5s.pt --img 640 --batch 1 --train
> 
> 转换正常（或至少看起来）并创建了模型（具有正确的类标签）。 ![图片](https://user-images.githubusercontent.com/3902189/128609746-68993c2e-45a3-40bf-aac1-62bf2492adfe.png)
> 
> 我已将它包含在 Apple 提供的用于采样对象检测的 Xcode 项目中 https://developer.apple.com/documentation/vision/recognizing_objects_in_live_capture ![图片](https://user-images.githubusercontent.com/3902189/128609787-9162b6a7-c7d3-4daa-b07a-e64947481c57.png)
> 
> 但是当项目在 iPhone 上运行时，它会抛出以下错误，并且在 iPhone 屏幕上什么都没有识别出来。 NSLocalizedDescription=神经网络中输出层'740'的大小与分类器中的类数不匹配。 ![图片](https://user-images.githubusercontent.com/3902189/128610083-51d68134-f86e-40f7-8851-d71dd1abd470.png)
> 
> 我还使用 Apple 在其网站上提供的 YOLOv3 MLModel 进行了检查，但效果很好：
> 
> ![图片](https://user-images.githubusercontent.com/3902189/128609708-cd572c0e-bd1b-44a4-85fc-3b17cfffe4b3.png)
> 
> 我的环境是：
> 
> MacOS Big Sur 11.5.1 Python 3.9 火炬 1.9.0 CPU onnx 1.10.0 scikit-learn 版本 0.24.2

How do you fix it finally?

**Comment by zjsjayce** [comment]
> I've exported the yolov5s.pt model to CoreML creating yolov5s.mlmodel. I've followed the instructions reported in #251 and #3238. The command I used is the following:
> 
> python3 export.py --weights yolov5s.pt --img 640 --batch 1 --train
> 
> The conversion has been ok (or at least it's seemed) and created the model (with the right class label). ![image](https://user-images.githubusercontent.com/3902189/128609746-68993c2e-45a3-40bf-aac1-62bf2492adfe.png)
> 
> I've included it in the Xcode project provided by Apple for sampling object detection https://developer.apple.com/documentation/vision/recognizing_objects_in_live_capture ![image](https://user-images.githubusercontent.com/3902189/128609787-9162b6a7-c7d3-4daa-b07a-e64947481c57.png)
> 
> But when the project runs on the iPhone, it throws the following error and nothing is recognised on the iPhone screen. NSLocalizedDescription=The size of the output layer '740' in the neural network does not match the number of classes in the classifier. ![image](https://user-images.githubusercontent.com/3902189/128610083-51d68134-f86e-40f7-8851-d71dd1abd470.png)
> 
> I've also done a check using the YOLOv3 MLModel provided by Apple in its site but it works perfectly:
> 
> ![image](https://user-images.githubusercontent.com/3902189/128609708-cd572c0e-bd1b-44a4-85fc-3b17cfffe4b3.png)
> 
> My environment is:
> 
> MacOS Big Sur 11.5.1 Python 3.9 torch 1.9.0 CPU onnx 1.10.0 scikit-learn version 0.24.2

How do you fix it finally?

**Comment by blocks-ai** [comment]
Hi friends,
any news ?

**Comment by blocks-ai** [comment]
@Kaisershmarren  how did you solved this problem ?

**Comment by glenn-jocher** [comment]
@blocks-ai, apologies for the delayed response. It seems like the error you are getting during the export process is due to the size of the output layer being different from the number of classes in the classifier. One possible solution to this issue would be to check whether the classes you are using in your classifier are matching the ones you have in your trained model.

Regarding the issue on iOS, unfortunately, I could not find enough information in your post that would help in debugging the problem. However, some possible ways to resolve this issue include checking to verify that the correct model file is being used in the iOS app, ensuring that the input size of the model matches the input image's size, and verifying the number of output layers and their sizes.

I hope this helps. If you have any further questions, please do not hesitate to ask.

**Comment by Kaisershmarren** [comment]
Hi all. Sorry for the late answer…
I managed to make yolov5 to work with apple vision by using what Leon0402 did in https://github.com/dbsystel/yolov5-coreml-tools.

I suggest you, @glenn-jocher, to use the source code provided by Leon, to integrate the new layers in your export logic for core ml. That’s because the object detection model in CoreML is useless if not compliant with vision requirements.

bye

**Comment by glenn-jocher** [comment]
@Kaisershmarren hello,

Thank you for letting us know about your success in using the source code provided by Leon0402 to make YOLOv5 work with Apple Vision. We appreciate your suggestion regarding integrating the new layers in our export logic for Core ML to ensure compliance with Vision requirements. 

We are always open to suggestions and contributions from the community, and we will definitely look into it. 

Have a great day!
